data/tutorials/language/0lg_10_polymorphic_variants.md

```diff
@@ -0,0 +1,713 @@
+
+The core type system of OCaml follows a [_nominal_](https://en.wikipedia.org/wiki/Nominal_type_system) discipline. Variants must be explicitly declared before being used. The typing discipline used for polymorphic variants and classes is different, as it is [_structural_](https://en.wikipedia.org/wiki/Structural_type_system).
+
+In the nominal approach of typing, types are first defined; later, when the type of an expression is inferred, three outcomes are possible:
+1. If a matching type is found, it becomes the inferred type.
+1. If any type can be applied, a type parameter is created.
+1. If typing inconsistencies are found, an error is raised.
+
+This is very similar to solving an equation in mathematics. Equations accept either zero, exactly one, several, or infinitely many numbers as solutions. Nominal type-checking finds that either zero, exactly one, or any type can be used in an expression.
+
+In the structural approach of typing, type definitions are optional, so they can be omitted. Type-checking an expression constructs a data structure that represents the types that are compatible with it. These data structures are displayed as type expressions sharing a resemblance with simple variants.
+
+<!--
+### Learning Goals
+
+The goal of this tutorial is to make the reader acquire the following capabilities:
+- Write polymorphic variants using code, from scratch, using mainstream features
+- Maintain existing code that is using polymorphic variants
+- Sort out polymorphic variant types and errors
+  - Add type annotation needed to resolve polymorphic variant type-checking issues
+  - Add or remove catch-all patterns in polymorphic variant pattern matching
+  - Insert or remove type casts
+- Refactor between simple and polymorphic variants
+- Choose to use polymorphic variants when really needed
+-->
+
+### A Note on Simple Variants and Polymorphism
+
+The type expression `'a list` does not designate a single type; it designates a family of types, basically all the types that can be created by substituting an actual type to type parameter `'a`. The type expressions `int list`, `bool option list`, or `(float -> float) list list` are real types. They're actual members of the _type family_ `'a list`. Types are intended to have inhabitants, type families don't.
+
+The identifiers `list`, `option`, and others are _type operators_. Just like functions, they take parameters. Although these parameters are not values, they are types. Their results aren't values but types, too.
+
+Simple variants are polymorphic, but not in the same sense as polymorphic variants.
+- Simple variants have [parametric polymorphism](https://en.wikipedia.org/wiki/Parametric_polymorphism)
+- Polymorphic variants have a form of [structural polymorphism](https://en.wikipedia.org/wiki/Structural_type_system)
+
+## A First Example
+
+### Creating Polymorphic Variant Types From Pattern Matching
```

Comment:
An even smaller example we could use here that I think would cement the intuition that polyvars are closer to the data itself than to defined types would be a list:

```ocaml
# let my_favorite_fruits = [ `banana ];;
val my_favorite_fruits : [> `banana ] list

# let my_favorite_fruits = [ `banana; `apple ];;
val my_favorite_fruits : [> `apple | `banana ] list

# let my_favorite_fruits = [ `banana; `apple; `kiwi ];;
val my_favorite_fruits : [> `apple | `banana | `kiwi ] list
```

Reply:
Could be a good path to build from:

1. `` `banana `` is of type `` [> `banana ] ``
2. if you put it together with other things that are their own type (`` `apple ``) then they form a larger type (`` [> `apple | `banana ] ``)